core/src/prediction/RegressionPredictionStrategy.cpp

```diff
@@ -100,11 +100,13 @@ PredictionValues RegressionPredictionStrategy::precompute_prediction_values(
     std::vector<double>& averages = values[i];
     averages.resize(1);
 
-    double average = 0.0;
+    double sum = 0.0;
+    double weight = 0.0;
     for (auto& sample : leaf_node) {
-      average += data->get_outcome(sample);
+      sum += data->get_weight(sample) * data->get_outcome(sample);
+      weight  += data->get_weight(sample);
```

Comment:
It's possible. I don't think any kind of bad numerical behavior can happen with positive weights, as the numerator will be roughly zero if the denominator is. I think the most natural solution is to add a positivity constraint to the validator in the R code.

Reply:
I thought about this more, and I think it makes the most sense to skip this leaf node if the total weight is too close to zero. This matches what we do when the leaf node is empty.

I like this behavior more than enforcing positivity, because I could see someone using an estimation procedure to determine sample weights. Some of the weights might end up being very close to zero, and it would be disruptive/ annoying if we rejected them.

Reply:
I meant a positivity check in the sense that we'd require that weights are > 0, so we wouldn't be rejecting weights for being small. If all weights are nonnegative, I don't think we'll have much trouble with floating point error with this, but I can also switch to a more stable algorithm, e.g. the one described in Section 4 [here](http://people.ds.cam.ac.uk/fanf2/hermes/doc/antiforgery/stats.pdf).

I have to push back a bit here on the idea of zeroing out leaves with small total weight. Inverse propensity weighting to minimize mean squared error when imputing missing observations, as in sample_weighting_examples/simple_regression_forest.R, involves weights that will often be near zero in regions in which we have plenty of data and can get a good regression estimate. A small weight w(x) just means that we think it's unlikely that we'll have to predict a value at x, so we don't want to go out of our way to get a good regression estimate there. But by ignoring low-weight leaves, we'd be ignoring the leaves that best estimate E[Y|X=x] at x where w(x') is small for x' near x.

Reply:
@jtibshirani has informed me that 'very close to zero' means equal to zero in the floating point sense, so we are in agreement. New version allow zeros weights and ignore leaves with total weight zero.